Adjusted code to be able to gather lost blocks/fulfill gaps

rylaix opened the pull request.

rylaix committed:
Adjusted code for cases, where gap fulfilling is needed.

rylaix committed:
Changed README to be Up to Date

fleupold commented on src/data_gathering.py:
```diff
@@ -188,18 +202,24 @@ def get_non_mev_transactions():
-    print(start_block, end_block)
+        # Get start and end blocks
+        latest_block_number = web3.eth.block_number
+        block_delay_seconds = config.get('block_delay_seconds', 0)
```
I think we should just assume the largest block number -1, we find in the Dune table is the latest block.

rylaix replied:
This logic was made Specially as it is, because Chen wanted it to be configurable (I mean, the delay).

The fulfilling gaps logic was not mentioned before, but is a good point to make it, so I decided to combine it with what we already have.
That's exactly why "fulfill_gap" Enables or disables the logic at all. In case of true, it bypasses the delay and offset (as logically, we need an exact gap).

Extract the latest block from dune, not from RPC (to be sure that is synced with Dune), that's a good point. (but not sure how we can do it, as we need to specify it via code. This can be done via SQL query itself but it will elevate the query cost)

fleupold replied:
How would someone want to run this script in the future? I think you would either want to run it for a specific block range or you would want to run it continuously (e.g. start off where you last were and index all new data available).

Having the concept of a `block_delay` is useful in the second case when we want to account for the time Dune takes to index all relevant transactions.

However, I'm also worried about the knowledge needed to correctly run/configure the script (5 values are too much).

If it's not possible to infer the most recent block dune has (by taking max block number - 1 from the `mevblocker.raw_bundles` table), I think we should at least reduce it to three

1.  start block (if not set, check if there is a last block stored in /data otherwise current block - 100)
2. end block (if not set, current block - block delay)
3. block delay

rylaix replied:
This approach seems good to me (If we do not want to have everything configurable via config itself).

> start block (if not set, check if there is a last block stored in /data otherwise current block - 100)


Basically, what you want to see is that our "start_block_offset" will disappear, to be a hardcoded logic .
I think it's okay, will do that.

fleupold reviewed (commented):
There are now 5 parameters that influence the range for which the data collection script runs (making it very complex to configure correctly)

- start_block_offset
- block_delay_seconds
- fulfill_gap
- start_block
- end_block

I think we can achieve the same with two:
- start_block (if not set, last processed block, if not present current block minus some reasonable value)
- end_block (if not set current block, and using the SQL results to figure out the last block that has actually been fully synced with dune)

Mefiseru commented:
Hello.
I think the implementation of Milestone 1 in the project just needs a final review. If it's everythink okay - I will start to commit more autonomously the Milestone 2 (If not, we will just lose days on answering).

fleupold commented:
> I think the implementation of Milestone 1 in the project just needs a final review.

Yes can you please merge outstanding work into the `main` branch so I can review it? This is required to consider the milestone as complete.

Mefiseru commented:
Yes of course. Will commit simplified logic and merge branch

rylaix committed:
Simplified configuration logic, and deleted unused "test.py"

rylaix committed:
Added new fields into SQL for Milestone 2 (See desc.)

They are commented at this moment, as I did not actually added these fields into Query on Dune. Will add them on face testing In Milestone 2 itself.

rylaix committed:
Enhanced logging logic to Show logs in console and Stderr.

Mefiseru commented:
Merged all new changes:)

Mefiseru commented:
@fleupold Hello! Just want to ensure, did you saw the last merged commits? Everything is okay?

fleupold commented:
Hey, yes looking at the script, things seem to be working fine and the data seems complete. I have three small points of improvement though

1. Please don't check in the `.env` file (it's too easy to accidentally leak secrets in a PR). Instead it's common to check in a `.env.example` file that people checking out the repo can copy to `.env` but `.env` itself should be part of the gitignore so it never makes it on github.
2. It's not clear to me what the purpose of the `confirmed_transactions_per_block` query is (it seems by default it does nothing). Since we are still fetching the content of the block from the RPC nodes (which is fine) it doesn't seem this query is useful for anything (it would be an alternative way of fetching if an RPC node isn't available, but I assume having one is anyway needed for step 2 so it's fine to not have this query at all)
3. The Dune queries that are part of the config are private and therefore not available from my API key. Can you either make them public or add to the README instructions on how to create your own version of the query?

Mefiseru commented:
Heya! I've seen that Dune just blocked my account (this is the reason why you we're not able to run queries). Dunno why this happened, did not got any advice or emails.
.env 018af55fd2e65be4fecec5334499924f9611eb40 (You're right about the process of env example, I forgot to implement it)
Readme 651a457c886bb8823493ecc821b01b2a501dabb4
Adjustments RPC logic gathering #9
